plugins/auth-backend-module-oidc-proxy-provider/src/helpers.ts

```diff
@@ -0,0 +1,113 @@
+/*
+ * Copyright 2024 The Backstage Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { LoggerService } from '@backstage/backend-plugin-api';
+import { AuthenticationError } from '@backstage/errors';
+import { JWK, KeyLike, importJWK, jwtVerify } from 'jose';
+import fetch from 'node-fetch';
+import { IDTokenInfo } from './types';
+
+export function createTokenValidator(
+  logger: LoggerService,
+  iss: string,
+  aud: string,
+  providedJwks?: (KeyLike | Uint8Array)[],
+): (token: string) => Promise<IDTokenInfo> {
+  let jwksUri: string = '';
+
+  return async function tokenValidator(token) {
+    // Perform discovery once to avoid two round trips on every authentication request.
+    if (jwksUri === '' && providedJwks === undefined) {
+      jwksUri = await discoverJwksUri(iss);
+    }
+
+    logger.info(`jwksUri: ${jwksUri}`);
+    // Verify the token was signed by the issuer.  Performs one round trip to
+    // the jwks uri every authentication request to fetch the current key set
+    // from the issuer.  May be optimized in the future, but this happens only
+    // at sign-in, not for every request to the backend.
+    //
+    // Refer to https://github.com/panva/jose/blob/v5.4.0/docs/functions/jwks_remote.createRemoteJWKSet.md#returns-1
```

Comment:
I'm still not entirely clear on why you didn't just use `createRemoteJWKSet` - that's what we do (passing the key set as the second argument to `jwtVerify`), and it keeps itself in sync perfectly as far as we know in the background, refreshing itself from the remote when needed etc.

Reply:
This code here is a reflection of me taking the code from the gcp-iap provider [here](https://github.com/jeffmccune/backstage/blob/305475c1b374cae3302fdfdad97c1be296e13af5/plugins/auth-backend-module-gcp-iap-provider/src/helpers.ts#L32) which uses the cloud sdk.

The cloud SDK, is GCP specific, but it's still OIDC under the hood, so I dove into _their_ source code to figure out how they verified the id token.

I then re-implemented what Google did natively here, which is why we ended up with this method.

In summary, an attempt on my part to remain faithful to gcp-iap, but perhaps missed the mark as soon as I started doing it the Google way instead of the other ways.

Reply:
Anyway, it works, so I'd rather not mess with it unless there's a compelling reason to.  Let me know.

Reply:
Ah, I recall now.  The other reason is the identity provider Holos uses is ZITADEL.  ZITADEL sometimes returns an empty jwks, `[]`, which caused `createRemoteJWKSet` to fail without grace, which in turn caused Backstage to fail with a mysterious user-facing error.

So, this is handles that edge case.

Refer to https://github.com/zitadel/zitadel/issues/6949#issuecomment-2035030448

Reply:
I added a comment to clarify the intent and why createRemoteJWKSet is avoided.